pytorch_lightning/trainer/lr_finder.py

```diff
@@ -0,0 +1,422 @@
+        self.auto_lr_find = False
+
+        # Initialize lr finder object (stores results)
+        lr_finder = _LRFinder(mode, min_lr, max_lr, num_training)
+
+        # Use special lr logger callback
+        callbacks = self.callbacks
+        self.callbacks = [_LRCallback(num_training, show_progress_bar=True)]
+
+        # No logging
+        logger = self.logger
+        self.logger = None
+
+        # Max step set to number of iterations
+        max_steps = self.max_steps
+        self.max_steps = num_training
+
+        # Disable standard progress bar for fit
+        progress_bar_refresh_rate = self.progress_bar_refresh_rate
+        self.progress_bar_refresh_rate = False
+
+        # Accumulation of gradients
+        accumulate_grad_batches = self.accumulate_grad_batches
+        self.accumulate_grad_batches = num_accumulation_steps
+
+        # Disable standard checkpoint
+        checkpoint_callback = self.checkpoint_callback
+        self.checkpoint_callback = False
+
+        # Dump model checkpoint
+        self._model_dump(save_path, model)
+
+        # Configure optimizer and scheduler
+        optimizers, _, _ = self.init_optimizers(model)
+
+        if len(optimizers) != 1:
+            raise MisconfigurationException(
+                f'`model.configure_optimizers()` returned {len(optimizers)}, but'
+                ' learning rate finder only works with single optimizer')
+        configure_optimizers = model.configure_optimizers
+        model.configure_optimizers = lr_finder._get_new_optimizer(optimizers[0])
+
+        # Fit, lr & loss logged in callback
+        self.fit(model, train_dataloader=train_dataloader)
+
+        # Promt if we stopped early
+        if self.global_step != num_training:
+            log.info('LR finder stopped early due to diverging loss.')
+
+        # Transfer results from callback to lr finder object
+        lr_finder.results.update({'lr': self.callbacks[0].lrs,
+                                  'loss': self.callbacks[0].losses})
+
+        # Finish by resetting variables so trainer is ready to fit model
+        self.auto_lr_find = auto_lr_find
+        self.logger = logger
+        self.callbacks = callbacks
+        self.max_steps = max_steps
+        self.progress_bar_refresh_rate = progress_bar_refresh_rate
+        self.accumulate_grad_batches = accumulate_grad_batches
+        self.checkpoint_callback = checkpoint_callback
+        model.configure_optimizers = configure_optimizers
+
+        # Reset model state
+        self._model_restore(save_path, model)
+        os.remove(save_path)
+
+        return lr_finder
+
+
+class _LRFinder(object):
+    """ LR finder object. This object stores the results of Trainer.lr_find().
+
+    Args:
+        mode: either `linear` or `exponential`, how to increase lr after each step
+
+        lr_min: lr to start search from
+
+        lr_max: lr to stop seach
+
+        num_iters: number of steps to take between lr_min and lr_max
```

Comment:
the `find_lr` method calls this `num_training` it'd be good to use the same variable name